### PR TITLE
Use single precision floating point numbers for coordinates

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: src/details/ArborX_DetailsNode.hpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ namespace ArborX
 {
 namespace Details
 {
-struct Node
+struct alignas(64) Node
```
Comment (dalg24) [inline src/details/ArborX_DetailsNode.hpp]: Provisory change.  My plan is to drop in the next PR along with `ptrdiff_t -> int` for storing children.

File: src/details/ArborX_Point.hpp [hunk 2]
```diff
@@ -23,42 +23,47 @@ class Point
...
+
 public:
   KOKKOS_INLINE_FUNCTION
   constexpr Point() noexcept = default;
 
   KOKKOS_INLINE_FUNCTION
-  constexpr Point(Data data)
-      : _data{std::move(data)}
+  constexpr Point(Abomination data)
+      : Point(data.xyz[0], data.xyz[1], data.xyz[2])
```
Comment (dalg24) [inline src/details/ArborX_Point.hpp]: Make sure you see that.  This allow the narrowing in aggregate initialization.  It is not pretty but that way I don't have to go change all occurrences yet.  I would prefer to defer fixing this to a later PR.

File: test/tstBoostGeometryAdapters.cpp [hunk 3]
```diff
@@ -75,8 +75,8 @@ BOOST_AUTO_TEST_CASE(distance)
 
   std::tie(a, b) = std::make_pair<details::Point, details::Point>(
       {{0., 0., 0.}}, {{1., 1., 1.}});
-  BOOST_TEST(details::distance(a, b) == std::sqrt(3.));
-  BOOST_TEST(bg::distance(a, b) == std::sqrt(3.));
+  BOOST_TEST(details::distance(a, b) == std::sqrt(3.f));
+  BOOST_TEST(bg::distance(a, b) == std::sqrt(3.0));
```
Comment (dalg24) [inline test/tstBoostGeometryAdapters.cpp]: Make sure you see that as well.  Boost promote the distance return type to double.

**Comment by dalg24** [comment]
Test this please

**Comment by dalg24** [comment]
... test this please

**Comment by aprokop** [comment]
[master.txt](https://github.com/arborx/ArborX/files/4287771/master.txt)
[single_precision.txt](https://github.com/arborx/ArborX/files/4287772/single_precision.txt)
```
$ compare_bench.py benchmarks master.json single_precision.json | grep median
Benchmark                                                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
BM_construction<ArborX::BVH<Serial>>/10000/0/manual_time_median                                  -0.0260         -0.0260          3028          2949          3028          2950
BM_construction<ArborX::BVH<Serial>>/100000/0/manual_time_median                                 -0.0391         -0.0391         30830         29626         30832         29626
BM_construction<ArborX::BVH<Serial>>/1000000/0/manual_time_median                                -0.0975         -0.0974        386772        349054        387582        349848
BM_construction<ArborX::BVH<Serial>>/10000/1/manual_time_median                                  -0.0300         -0.0300          3102          3009          3103          3010
BM_construction<ArborX::BVH<Serial>>/100000/1/manual_time_median                                 -0.0435         -0.0435         32450         31039         32452         31041
BM_construction<ArborX::BVH<Serial>>/1000000/1/manual_time_median                                -0.0896         -0.0894        409494        372818        410257        373590
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/0/2/manual_time_median                         +0.0670         +0.0670         43676         46601         43680         46605
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/0/2/manual_time_median                       +0.0587         +0.0587        466609        494020        466632        494043
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/2/manual_time_median                     +0.0434         +0.0434       5004970       5222242       5005387       5222639
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/1/3/manual_time_median                         +0.0526         +0.0526         44631         46979         44635         46983
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/1/3/manual_time_median                       +0.0394         +0.0394        631491        656365        631520        656387
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/1/3/manual_time_median                     +0.0237         +0.0237       9005085       9218627       9005622       9219163
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/0/2/manual_time_median                    +0.0393         +0.0393         66228         68828         66233         68833
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/0/2/manual_time_median                  +0.0383         +0.0383        697268        723973        697296        724006
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/0/2/manual_time_median                +0.0349         +0.0349       7314117       7569551       7314650       7570140
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/1/3/manual_time_median                    +0.0375         +0.0375         15825         16420         15828         16422
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/1/3/manual_time_median                  +0.0366         +0.0366        111013        115080        111019        115086
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/1/3/manual_time_median                +0.0286         +0.0285        993571       1021944        993596       1021963
BM_construction<ArborX::BVH<OpenMP>>/10000/0/manual_time_median                                  -0.0274         -0.0274          3072          2988          3074          2990
BM_construction<ArborX::BVH<OpenMP>>/100000/0/manual_time_median                                 -0.0424         -0.0424         31442         30109         31445         30113
BM_construction<ArborX::BVH<OpenMP>>/1000000/0/manual_time_median                                -0.1002         -0.1000        390838        351691        391604        352459
BM_construction<ArborX::BVH<OpenMP>>/10000/1/manual_time_median                                  -0.0278         -0.0278          3134          3047          3136          3049
BM_construction<ArborX::BVH<OpenMP>>/100000/1/manual_time_median                                 -0.0387         -0.0387         32755         31487         32758         31490
BM_construction<ArborX::BVH<OpenMP>>/1000000/1/manual_time_median                                -0.0878         -0.0876        408977        373077        409749        373847
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/2/manual_time_median                         +0.0677         +0.0677         43550         46497         43554         46501
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/2/manual_time_median                       +0.0624         +0.0624        464219        493199        464240        493221
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/2/manual_time_median                     +0.0500         +0.0501       4966108       5214594       4966216       5214998
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/1/3/manual_time_median                         +0.0633         +0.0633         44436         47249         44439         47252
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/1/3/manual_time_median                       +0.0554         +0.0554        627167        661921        627187        661947
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/1/3/manual_time_median                     +0.0417         +0.0417       8943809       9316931       8944381       9317473
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/0/2/manual_time_median                    +0.0441         +0.0441         66199         69119         66204         69123
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/0/2/manual_time_median                  +0.0449         +0.0449        695543        726779        695571        726810
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/0/2/manual_time_median                +0.0455         +0.0455       7277203       7608026       7277729       7608610
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/1/3/manual_time_median                    +0.0414         +0.0414         15878         16536         15881         16539
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/1/3/manual_time_median                  +0.0439         +0.0439        110891        115762        110896        115769
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/1/3/manual_time_median                +0.0439         +0.0439        986864       1030209        986895       1030250
BM_construction<ArborX::BVH<Cuda>>/10000/0/manual_time_median                                    +0.0031         +0.0023          1140          1144          1425          1428
BM_construction<ArborX::BVH<Cuda>>/100000/0/manual_time_median                                   -0.0293         -0.0262          2854          2771          3155          3072
BM_construction<ArborX::BVH<Cuda>>/1000000/0/manual_time_median                                  +0.1726         +0.1591          7765          9105          8263          9577
BM_construction<ArborX::BVH<Cuda>>/10000/1/manual_time_median                                    +0.0025         +0.0021          1135          1138          1420          1423
BM_construction<ArborX::BVH<Cuda>>/100000/1/manual_time_median                                   -0.0302         -0.0273          2862          2775          3162          3076
BM_construction<ArborX::BVH<Cuda>>/1000000/1/manual_time_median                                  +0.1675         +0.1576          7751          9049          8249          9549
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/0/2/manual_time_median                           -0.0084         -0.0088          1968          1952          2051          2032
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/0/2/manual_time_median                         -0.2190         -0.2130         12599          9840         12956         10196
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/2/manual_time_median                       -0.1697         -0.1687        118250         98185        118945         98881
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/1/3/manual_time_median                           -0.0227         -0.0224          2158          2109          2240          2190
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/1/3/manual_time_median                         -0.2789         -0.2735         18026         12998         18382         13354
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/3/manual_time_median                       -0.1848         -0.1842        244975        199710        245667        200408
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/2/manual_time_median                      -0.0200         -0.0190          1044          1023          1123          1102
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/2/manual_time_median                    -0.0351         -0.0333          7654          7385          8013          7746
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/2/manual_time_median                  +0.2200         +0.2173         57482         70131         58174         70815
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/1/3/manual_time_median                      +0.0185         +0.0171           961           979          1041          1058
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/1/3/manual_time_median                    -0.0513         -0.0460          3316          3146          3669          3501
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/1/3/manual_time_median                  -0.0523         -0.0495         10673         10115         11312         10753
```

File: src/details/ArborX_DetailsNode.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: > I would like to consider whether that should be a configurable parameter.

What does "that" refer to?

Comment (aprokop) [reply src/details/ArborX_DetailsNode.hpp]: Whether we are storing children as size_t/ptrdiff_t, or int.

Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: I'd say we can fix later.  One may ask to use double precision as well ;)

Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: There is a lot more involved for making things work with more than 2B objects.
For instance we explicitly take indices as `int` in our lambdas/functions.  Also default index types would need to change in execution policies.

**Comment by aprokop** [comment]
Changing `Point` to `alignas(16)` slowed things down. Changing `Box` to `align(32)` sped things up.

[single_precision_Box32.txt](https://github.com/arborx/ArborX/files/4288787/single_precision_Box32.txt)

```
 $ compare_bench.py benchmarks single_precision.json single_precision_Box32.json | grep median    
Benchmark                                                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------        
BM_construction<ArborX::BVH<Serial>>/10000/0/manual_time_median                                  +0.0283         +0.0283          2949          3032          2950          3033
BM_construction<ArborX::BVH<Serial>>/100000/0/manual_time_median                                 +0.0324         +0.0324         29626         30586         29626         30587
BM_construction<ArborX::BVH<Serial>>/1000000/0/manual_time_median                                +0.0310         +0.0309        349054        359877        349848        360650
BM_construction<ArborX::BVH<Serial>>/10000/1/manual_time_median                                  +0.0276         +0.0273          3009          3092          3010          3092
BM_construction<ArborX::BVH<Serial>>/100000/1/manual_time_median                                 +0.0289         +0.0287         31039         31937         31041         31931
BM_construction<ArborX::BVH<Serial>>/1000000/1/manual_time_median                                +0.0258         +0.0254        372818        382455        373590        383096
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/0/2/manual_time_median                         +0.0004         +0.0002         46601         46619         46605         46613
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/0/2/manual_time_median                       +0.0031         +0.0029        494020        495550        494043        495485
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/2/manual_time_median                     +0.0072         +0.0072       5222242       5259841       5222639       5260283
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/1/3/manual_time_median                         +0.0067         +0.0067         46979         47294         46983         47298
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/1/3/manual_time_median                       +0.0005         +0.0005        656365        656666        656387        656692
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/1/3/manual_time_median                     +0.0015         +0.0015       9218627       9232223       9219163       9232754
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/0/2/manual_time_median                    +0.0023         +0.0023         68828         68984         68833         68988
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/0/2/manual_time_median                  +0.0008         +0.0008        723973        724555        724006        724580
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/0/2/manual_time_median                +0.0016         +0.0016       7569551       7581397       7570140       7581936
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/1/3/manual_time_median                    -0.0032         -0.0032         16420         16368         16422         16370
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/1/3/manual_time_median                  -0.0032         -0.0031        115080        114717        115086        114724
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/1/3/manual_time_median                +0.0001         +0.0001       1021944       1022016       1021963       1022050
BM_construction<ArborX::BVH<OpenMP>>/10000/0/manual_time_median                                  +0.0392         +0.0391          2988          3105          2990          3107
BM_construction<ArborX::BVH<OpenMP>>/100000/0/manual_time_median                                 +0.0390         +0.0389         30109         31282         30113         31285
BM_construction<ArborX::BVH<OpenMP>>/1000000/0/manual_time_median                                +0.0426         +0.0425        351691        366678        352459        367453
BM_construction<ArborX::BVH<OpenMP>>/10000/1/manual_time_median                                  +0.0395         +0.0395          3047          3167          3049          3169
BM_construction<ArborX::BVH<OpenMP>>/100000/1/manual_time_median                                 +0.0365         +0.0365         31487         32636         31490         32639
BM_construction<ArborX::BVH<OpenMP>>/1000000/1/manual_time_median                                +0.0450         +0.0450        373077        389881        373847        390665
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/2/manual_time_median                         -0.0022         -0.0022         46497         46397         46501         46400
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/2/manual_time_median                       -0.0026         -0.0027        493199        491894        493221        491913
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/2/manual_time_median                     -0.0005         -0.0005       5214594       5212072       5214998       5212472
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/1/3/manual_time_median                         -0.0005         -0.0005         47249         47224         47252         47227
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/1/3/manual_time_median                       -0.0021         -0.0021        661921        660536        661947        660562
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/1/3/manual_time_median                     -0.0019         -0.0019       9316931       9299552       9317473       9300075
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/0/2/manual_time_median                    -0.0129         -0.0129         69119         68228         69123         68232
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/0/2/manual_time_median                  -0.0130         -0.0130        726779        717319        726810        717344
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/0/2/manual_time_median                -0.0108         -0.0108       7608026       7525812       7608610       7526353
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/1/3/manual_time_median                    -0.0113         -0.0113         16536         16350         16539         16352
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/1/3/manual_time_median                  -0.0090         -0.0090        115762        114719        115769        114725
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/1/3/manual_time_median                -0.0087         -0.0087       1030209       1021222       1030250       1021254
BM_construction<ArborX::BVH<Cuda>>/10000/0/manual_time_median                                    +0.0017         +0.0015          1144          1146          1428          1430
BM_construction<ArborX::BVH<Cuda>>/100000/0/manual_time_median                                   +0.0435         +0.0396          2771          2891          3072          3194
BM_construction<ArborX::BVH<Cuda>>/1000000/0/manual_time_median                                  -0.0800         -0.0734          9105          8377          9577          8874
BM_construction<ArborX::BVH<Cuda>>/10000/1/manual_time_median                                    -0.0021         -0.0017          1138          1135          1423          1420
BM_construction<ArborX::BVH<Cuda>>/100000/1/manual_time_median                                   +0.0426         +0.0389          2775          2893          3076          3195
BM_construction<ArborX::BVH<Cuda>>/1000000/1/manual_time_median                                  -0.0746         -0.0712          9049          8374          9549          8869
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/0/2/manual_time_median                           -0.0119         -0.0123          1952          1929          2032          2007
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/0/2/manual_time_median                         -0.0031         -0.0031          9840          9809         10196         10164
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/2/manual_time_median                       +0.0063         +0.0062         98185         98800         98881         99495
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/1/3/manual_time_median                           -0.0039         -0.0048          2109          2101          2190          2179
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/1/3/manual_time_median                         +0.0051         +0.0049         12998         13064         13354         13420
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/3/manual_time_median                       +0.0154         +0.0154        199710        202790        200408        203490
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/2/manual_time_median                      -0.0192         -0.0197          1023          1003          1102          1080
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/2/manual_time_median                    -0.0280         -0.0266          7385          7178          7746          7540
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/2/manual_time_median                  -0.0472         -0.0465         70131         66824         70815         67520
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/1/3/manual_time_median                      -0.0107         -0.0106           979           969          1058          1047
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/1/3/manual_time_median                    -0.0155         -0.0139          3146          3097          3501          3452
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/1/3/manual_time_median                  +0.0047         +0.0003         10115         10163         10753         10756

```

I also think we need to update `bvh_driver` to operate on both `Point` and `Box` so that we can test both. Both cases are important, but right now we are only testing boxes.

**Comment by aprokop** [comment]
Investigating performance issues with single precision patch.

1. CUDA radius_search
   `BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/2/manual_time_median`
   `+0.0594         +0.0589         57482         60894`

   Comparison of Kokkos timers indicates that the main issue is in the  pass (3.47 -> 3.78). That is likely a combination of  access with random access through permutations.

2. CUDA construction
   `BM_construction<ArborX::BVH<Cuda>>/10000/1/manual_time_median`
   `+0.2371         -0.0005          1135          1404`

   The main problem is coming from `sort_morton_codes_and_init_leaves` (2.55 -> 3.72). However, ArborX_initialize_leaf_nodes` kernel is fine (0.10 -> 0.07). Thus, must be the sorting. However, sorting is independent of float. So, WTF???

3. OpenMP radius_search
   `BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/0/2`
   `+0.0546         +0.0546       7277203       7674254`

   Second pass is the problem.

4. OpenMP knn search
   `BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/2`
   `+0.0518         +0.0519       4966108       5223357`

   `ArborX_perform_nearest_queries` is the problem

Overall, it seems that it is the interaction between random access and
not aligned data that causes most issues. The remaining question is what happens in 2.

**Comment by aprokop** [comment]
To see the effect of permute on non-aligned data, I ran a hacked version where queries are not permuted (
[patch](https://github.com/arborx/ArborX/files/4289818/no_permute_patch.txt). Both master and this branch were patched. Here are the results:
[master_no_permute.txt](https://github.com/arborx/ArborX/files/4289851/master_no_permute.txt)
[single_precision_int_no_permute.txt](https://github.com/arborx/ArborX/files/4289854/single_precision_int_no_permute.txt)

```
$ compare_bench.py benchmarks master_no_permute.json single_precision_int_no_permute.json  | grep median | grep search
Benchmark                                                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------        
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/0/2/manual_time_median                         -0.0165         -0.0165         47106         46331         47110         46335
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/0/2/manual_time_median                       -0.0196         -0.0196        513898        503821        513923        503844
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/2/manual_time_median                     -0.1074         -0.1074       6480868       5785017       6481234       5785386
BM_knn_search<ArborX::BVH<Serial>>/10000/10000/10/1/3/manual_time_median                         -0.0154         -0.0154         52072         51270         52076         51274
BM_knn_search<ArborX::BVH<Serial>>/100000/100000/10/1/3/manual_time_median                       -0.0145         -0.0145        752932        741984        752959        742013
BM_knn_search<ArborX::BVH<Serial>>/1000000/1000000/10/1/3/manual_time_median                     -0.0705         -0.0705      11782743      10951747      11783325      10952219
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/0/2/manual_time_median                    +0.0190         +0.0190         72428         73807         72433         73811
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/0/2/manual_time_median                  +0.0146         +0.0146        790425        801985        790460        802019
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/0/2/manual_time_median                -0.0820         -0.0820      10093404       9266155      10093974       9266723
BM_radius_search<ArborX::BVH<Serial>>/10000/10000/10/0/1/3/manual_time_median                    +0.0254         +0.0256         21407         21951         21406         21954
BM_radius_search<ArborX::BVH<Serial>>/100000/100000/10/0/1/3/manual_time_median                  +0.0154         +0.0153        147703        149972        147714        149980
BM_radius_search<ArborX::BVH<Serial>>/1000000/1000000/10/0/1/3/manual_time_median                +0.0336         +0.0336       1128373       1166304       1128410       1166339
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/2/manual_time_median                         -0.0236         -0.0236         47323         46207         47327         46210
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/2/manual_time_median                       -0.0258         -0.0258        515901        502595        515918        502615
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/2/manual_time_median                     -0.1111         -0.1111       6493121       5772026       6493488       5772383
BM_knn_search<ArborX::BVH<OpenMP>>/10000/10000/10/1/3/manual_time_median                         -0.0192         -0.0192         52114         51115         52118         51118
BM_knn_search<ArborX::BVH<OpenMP>>/100000/100000/10/1/3/manual_time_median                       -0.0206         -0.0206        754793        739207        754822        739236
BM_knn_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/1/3/manual_time_median                     -0.0772         -0.0772      11831346      10917984      11831791      10918434
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/0/2/manual_time_median                    +0.0223         +0.0223         72493         74107         72497         74111
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/0/2/manual_time_median                  +0.0152         +0.0152        792581        804656        792615        804690
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/0/2/manual_time_median                -0.0781         -0.0781      10076272       9289762      10076858       9290321
BM_radius_search<ArborX::BVH<OpenMP>>/10000/10000/10/0/1/3/manual_time_median                    +0.0430         +0.0430         21162         22073         21165         22075
BM_radius_search<ArborX::BVH<OpenMP>>/100000/100000/10/0/1/3/manual_time_median                  +0.0316         +0.0316        146058        150673        146064        150681
BM_radius_search<ArborX::BVH<OpenMP>>/1000000/1000000/10/0/1/3/manual_time_median                +0.0487         +0.0487       1116885       1171241       1116926       1171277
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/0/2/manual_time_median                           -0.5012         -0.4764          1460           728          1541           807
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/0/2/manual_time_median                         -0.2444         -0.2534         11795          8912         12417          9271
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/2/manual_time_median                       -0.2473         -0.2460        130281         98058        130973         98757
BM_knn_search<ArborX::BVH<Cuda>>/10000/10000/10/1/3/manual_time_median                           -0.4735         -0.4528          1646           866          1727           945
BM_knn_search<ArborX::BVH<Cuda>>/100000/100000/10/1/3/manual_time_median                         -0.2195         -0.2248         23423         18281         24046         18640
BM_knn_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/3/manual_time_median                       -0.1308         -0.1305        342271        297510        342976        298215
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/2/manual_time_median                      -0.1217         -0.1071           573           503           652           582
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/2/manual_time_median                    -0.1926         -0.2118          6948          5610          7578          5973
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/2/manual_time_median                  -0.0101         -0.0100        112486        111354        113194        112061
BM_radius_search<ArborX::BVH<Cuda>>/10000/10000/10/0/1/3/manual_time_median                      -0.0548         -0.0455           447           422           525           501
BM_radius_search<ArborX::BVH<Cuda>>/100000/100000/10/0/1/3/manual_time_median                    -0.0758         -0.0644          2197          2030          2551          2387
BM_radius_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/1/3/manual_time_median                  +0.0104         +0.0097         11382         11500         12024         12141
```
Indeed, without permute this PR is almost unconditionally better (except very slight increase in spatial search for Serial and OpenMP). Overall, the main beneficiary of the single precision seems to be knn_search for CUDA.

**Comment by aprokop** [comment]
And just to note: Morton calculation function still takes in `double`. Switching to `float` made no difference there.

**Comment by aprokop** [comment]
I'm not holding this but I'm also not ready to approve. If you want to merge it, go ahead.

**Comment by aprokop** [comment]
One other issue that we have not addressed in this PR is that narrowing user `double` data to `float` may result in a situation where original data intersects and our data does not. This seems like an extreme user case, and we are not going to worry about it.

**Pull request merged**